plugins/modules/foreman_snapshot.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# (c) 2019 Manisha Singhal (ATIX AG)
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: foreman_snapshot
+short_description: Manage Foreman Snapshots
+description:
+  - "Manage Foreman Snapshots for Host Entities"
+  - "This beta version can create, update, revert and delete snapshots from preexisting host"
+author:
+  - "Manisha Singhal (@Manisha15) ATIX AG"
+requirements:
+    - "apypie >= 0.0.1"
+options:
+  server_url:
+    description:
+      - URL of Foreman server
+    required: true
+  username:
+    description:
+      - Username on Foreman server
+    required: true
+  password:
+    description:
+      - Password for user accessing Foreman server
+    required: true
+  validate_certs:
+    aliases: [ verify_ssl ]
+    description:
+      - Verify SSL of the Foreman server
+    required: false
+    default: true
+    type: bool
+  name:
+    description:
+      - Name of Snapshot
+    required: true
+  description:
+    description:
+      - Description of Snapshot
+    required: false
+  host:
+    description:
+      - Name of related Host
+    required: true
+
+'''
+
+EXAMPLES = '''
+- name: "Create a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    state: present
+
+- name: "Update a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    description: "description"
+    state: present
+
+- name: "Revert a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    state: reverted
+
+- name: "Delete a Snapshot"
+  foreman_snapshot:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    name: "new_host"
+    host: host_name
+    state: absent
+'''
+
+RETURN = ''' # '''
+
+
+from ansible.module_utils.foreman_helper import ForemanEntityApypieAnsibleModule
+
+
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'description': 'description',
+    'host_id': 'host_id',
```

Comment:
i'd prefer this to be `'host': 'host_id'` and then overriding the value in L136, otherwise `snapshot_dict` ends up with having both, `host` and `host_id`.

Reply:
This mapping (the one you suggested ) creates problem while updating the snapshot. Unfortunately during sanitizing entity in update part of ensure_resource it can't identity host_id mapping.

Reply:
interesting. the snapshot plugin is really weird.